esp-hal/src/timer/systimer.rs

```diff
@@ -629,65 +629,65 @@ mod asynch {
 
     impl<'a> AlarmFuture<'a> {
         pub(crate) fn new(alarm: &'a Alarm) -> Self {
+            // For some reason, on the S2 we need to enable the interrupt before we
+            // read its status. Doing so in the other order causes the interrupt
+            // request to never be fired.
             alarm.enable_interrupt(true);
-
             Self { alarm }
         }
 
-        fn event_bit_is_clear(&self) -> bool {
-            SYSTIMER::regs()
-                .int_ena()
-                .read()
-                .target(self.alarm.channel())
-                .bit_is_clear()
+        fn is_done(&self) -> bool {
+            self.alarm.is_interrupt_set()
         }
     }
 
     impl core::future::Future for AlarmFuture<'_> {
         type Output = ();
 
         fn poll(self: Pin<&mut Self>, ctx: &mut Context<'_>) -> Poll<Self::Output> {
+            // Interrupts are enabled, so we need to register the waker before we check for
+            // done. Otherwise we might miss the interrupt that would wake us.
             WAKERS[self.alarm.channel() as usize].register(ctx.waker());
 
-            if self.event_bit_is_clear() {
+            if self.is_done() {
                 Poll::Ready(())
             } else {
                 Poll::Pending
             }
         }
     }
 
+    impl Drop for AlarmFuture<'_> {
+        fn drop(&mut self) {
+            self.alarm.enable_interrupt(false);
+            self.alarm.clear_interrupt();
```

Comment:
Clearing the interrupt here breaks cancellation I think, you ideally want to be able to re-create this future in the future for the same alarm.

Reply:
Ah nvm, I don't think the cancellation semantics where ever decided on the timers.

Reply:
Yeah this future isn't cancellation-safe. Neither does it have to be, either.

Reply:
Still, I removed this clear in the last commit, where I also rework the async wait a bit.